GoReleaser build and release support

kskewes opened the pull request:
GoReleaser build and release support

Simplified build (and release?) per #134.
Not automatically hooked into Travis or anything (yet).

**Binaries tested with `--version`:**
- linux (docker images as well)
- windows

**Binaries not tested:**
- darwin (don't have mac)

**Build and Publish tested:**
- Create [GitHub Token](https://goreleaser.com/environment/)
- In `.goreleaser.yml` comment out `release.disable: true`
- `git commit`
- `git tag v0.7.1 && git push`
- `make publish`
- single archive per platform contains both `controller` and `kubeseal` binaries, plus kubernetes yaml's
- locally docker images are created

**Not Tested:**
- docker image push - In `.goreleaser.yml` comment out `docker.skip_push: true`

**Build fork/dirty/etc:** (assuming merged, patch on upstream or blog link below)
```
make snapshot
```
This results in binaries, archives and docker images tagged with full git commit hash.

**To Do:**
- [x] TODO: comment added for brew to change to [published formula](https://github.com/Homebrew/homebrew-core/blob/master/Formula/kubeseal.rb)
- [x] confirm ok publishing archive instead of binaries
- [x] yamls added to archive
- [x] DOCS update and simplification

First time using goreleaser so... :)

kskewes commented on .goreleaser.yml:
```diff
@@ -0,0 +1,102 @@
+  - bitnami-labs/sealed-secrets-controller:amd64-{{ .Version }}
+  # Comment below out to push to configured repository
+  skip_push: true
+- goarch: arm64
```
Ahh, good point. Although it was copying in the correct binaries, any shell commands run by the controller would have ELF exec failed.

Updated Dockerfile with `ARCH` build arg. 
Defaults to amd64 and verified can still build it without goreleaser.
`ARCH` gets overwritten by goreleaser in arm64 and armv6 docker image sections.

```
$ docker image inspect bitnami-labs/sealed-secrets-controller:{amd64,arm64,armv6}-latest | grep Arch
        "Architecture": "amd64",
        "Architecture": "arm64",
        "Architecture": "arm",
```

Testing arm64 build:
```
rock64@k8s-w-01:~$ uname -m
aarch64
rock64@k8s-w-01:~$ sudo docker run bitnami-labs/sealed-secrets-controller:arm64-latest --version
controller version: a31f1ee3da9c5972fd865761910e3876e665e17b
rock64@k8s-w-01:~$ sudo docker inspect bitnami-labs/sealed-secrets-controller:arm64-latest | grep Arch
        "Architecture": "arm64",
```

WDYT?

kskewes commented on .goreleaser.yml:
```diff
@@ -0,0 +1,102 @@
+  homepage: "https://github.com/bitnami-labs/sealed-secrets/"
+  description: "A Kubernetes controller and tool for one-way encrypted Secrets"
+  skip_upload: true
+  # Replace with existing brew formula
```
Sounds good to me. I don't have a Mac, so best for someone familiar with it pick up later. 
Have commented out, but can delete if preferable.

kskewes commented on .goreleaser.yml:
```diff
@@ -0,0 +1,102 @@
+  # Comment below out to push to configured repository
+  skip_push: true
+
+archive:
```
It would seem GoReleaser doesn't (yet?) support selecting which binaries from the `builds:` section get included in archives.
There's a ticket about it here: https://github.com/goreleaser/goreleaser/issues/705
Seems like a fairly major arch change in GoReleaser required.

On this topic, are you happy with the default inclusion of README and LICENSE files? 
We can exclude these if you prefer.

[GoReleaser Archive docs](https://goreleaser.com/customization/#Archive)

anguslees replied:
> On this topic, are you happy with the default inclusion of README and LICENSE files?
We can exclude these if you prefer.

Yep, including those files is fine and appropriate.

kskewes commented on .goreleaser.yml:
```diff
@@ -0,0 +1,102 @@
+  skip_push: true
+
+archive:
+  replacements:
```
Thanks, good point.

> Also (note to self): This will change the install instructions. We need to time an update to the README to uncompress the new tarball. Perhaps we just link to github releases and don't give cut+paste commands.

Couldn't help but think through this a bit. 

Perhaps we can simplify the README installation process right down if:
- commit `controller.yaml`'s to [/examples](https://github.com/bitnami-labs/sealed-secrets/tree/master/examples) with `latest` image. Serious users will pin versions/build/etc.
- remove/update `archive.replacements`. Not sure about arm64/aarch64 or armv7/armv6.
- CRD assumption, fold into controller.yaml (with kind: namespace, etc), TPR separate. 
- move Source/Test/Developing to separate CONTRIBUTING.md or DEVELOPING.md?

## Server Installation

```
# For clusters with RBAC
wget https://github.com/bitnami-labs/sealed-secrets/examples/controller.yaml
kubectl apply -f controller.yaml

# For clusters without RBAC
wget https://github.com/bitnami-labs/sealed-secrets/examples/controller-norbac.yaml
kubectl apply -f controller-norbac.yaml
```

## Client Installation
MacOS: 
```
brew install kubeseal
```
Linux:
```
# get latest release version
release=$(curl --silent "https://api.github.com/repos/bitnami-labs/sealed-secrets/releases/latest" | sed -n 's/.*"tag_name": *"\([^"]*\)".*/\1/p')

# Download release
wget https://github.com/bitnami-labs/sealed-secrets/releases/download/$release/sealed-secrets_$release_$(uname)_$(uname -m).tar.gz

# Extract release
tar -zxvf sealed-secrets_$release_$(uname)_$(uname -m).tar.gz

# Install in path (if required)
sudo install -m 755 kubeseal /usr/local/bin/kubeseal
```

Windows:
- download latest [release](https://github.com/bitnami-labs/sealed-secrets/releases/latest)
- extract archive

## Client usage 

(Existing docs is fine. I would just add fetching cert here)

### Optional one time process - save encryption cert to file

The public cert can be commited to git and supplied as an argument to `kubeseal` instead of requiring cluster access. For example: `kubeseal --cert mycert.pem`
```
# Save public cert to file (accessible controller)
kubeseal --fetch-cert > mycert.pem

# Save public cert to a file (kubectl access to secret)
kubectl get secret -n sealed-secrets sealed-secrets-key -o json | jq -r '.data | ."tls.crt"' > mycert.pem
```
### Creating a Sealed Secret

.... from [existing docs](https://github.com/bitnami-labs/sealed-secrets#usage)

kskewes commented:
Thanks very much for reviewing Gus.

kskewes commented:
Have had a go at docs tidy up to go with `goreleaser` use.
Easiest to preview here on my fork here: https://github.com/kskewes/sealed-secrets/tree/goreleaser

Specifically:
- [README.md](https://github.com/kskewes/sealed-secrets/blob/goreleaser/README.md)
- [CONTRIBUTING.md](https://github.com/kskewes/sealed-secrets/blob/goreleaser/CONTRIBUTING.md)
- [MIGRATION.md - MIGRATING-FROM-TPR.md?](https://github.com/kskewes/sealed-secrets/blob/goreleaser/MIGRATION.md)

TBD:
- `controller.yaml` generation from jsonnet - update Makefile and store in `/examples/` ?
- CONTRIBUTING and MIGRATION updates for goreleaser based install flow.

kskewes commented:
Seems pretty complete. Happy to amend as needed.
Goreleaser doesn't generate yaml from jsonnet hence add `/examples/controller*.yaml` with no image tag (`:latest`).
Follow up task to enable release pushing to github & docker when you're ready, or I can set like that as part of this PR.
Haven't touched Makefile.

anguslees commented:
I think this is mostly awesome, and ready to merge (with the above comments).

You haven't hooked this into anything that actually invokes `goreleaser` yet, so nothing will change immediately as part of this PR (good).  I think we need to wrap invoking `goreleaser` with something that generates versioned k8s manifests, and pushes those manifests to github release artefacts.  Easiest to do that after we have some experience with driving the tool/workflow manually.

kskewes commented on README.md:
```diff
@@ -11,165 +11,92 @@ decrypted only by the controller running in the target cluster and
-# Install server-side controller into kube-system namespace (by default)
-$ kubectl create -f https://github.com/bitnami-labs/sealed-secrets/releases/download/$release/controller.yaml
+# For clusters with RBAC
+$ wget https://github.com/bitnami-labs/sealed-secrets/examples/controller.yaml
```
Okay, sure will try that and verify flow with new tag in my fork.

kskewes replied:
Generating and adding YAML's to archive with `builds.hooks.post` was straight forward with two exceptions:
1. `Makefile` target builds docker image as part of generating YAML, this duplicates goreleaser and could lead to different docker images, see [proposed change](https://github.com/bitnami-labs/sealed-secrets/pull/140/files#diff-b67911656ef5d18c4ae36cb6741b7965R61)
2. `Makefile` adds `+dirty` to `$(VERSION)` if repo dirty but `+` is invalid image tag character, see [proposed change](https://github.com/bitnami-labs/sealed-secrets/pull/140/files#diff-b67911656ef5d18c4ae36cb6741b7965R10)
I'm trying not to alter existing flow too much but there is overlapping functionality.

RE: Publishing to GitHub as part of Release:
Looks like the `Release` struct in [Goreleaser](https://github.com/goreleaser/goreleaser/blob/master/pkg/config/config.go#L159) doesn't have a `Files` var like some of the others.
I've asked on their Slack channel about it. Imagine it would take a bit of plumbing to wire it through if they are open to a PR. Maybe there's an alternate method?

kskewes commented on README.md:
````diff
@@ -11,165 +11,92 @@ decrypted only by the controller running in the target cluster and
-ensure you are using the correct file.
+```sh
+# Save public cert to a file (kubectl access to secret)
+$ kubectl get secret -n kube-system sealed-secrets-key -o json | jq -r '.data | ."tls.crt"' | base64 -d > mycert.pem
````
Maybe not. 
When I went through our installation across a few clusters I chose not to expose the controller due to associated networking and auth. Instead just pull cert from k8s and commit into our private repo for use by those with repo rights (developers).
Guess other option is to direct to controller logs for it as mentioned elsewhere.
Will remove.

kskewes commented:
- Added `Makefile` targets to wrap typical `goreleaser` commands. 
  Actual publishing to GitHub and docker push still prohibited by couple lines in `.goreleaser.yml`, please see PR message at top.

- YAML's get generated and added to archive now. [Details here](https://github.com/bitnami-labs/sealed-secrets/pull/140#discussion_r259667305)

kskewes commented:
Rebased on master with TPR removed.

kskewes commented:
Yes! Thanks for your patience.

anguslees commented:
Merging!

bors r+
